ossfuzz/aspell_fuzzer.cpp

```diff
@@ -0,0 +1,199 @@
+  {
+    FUZZ_DEBUG("Configuration parsing failed");
+    goto EXIT_LABEL;
+  }
+
+  // Move the data pointer past the config.
+  data_str += rc;
+  size -= rc;
+
+  FUZZ_DEBUG("Document: %.*s", (int)size, data_str);
+
+  // Convert the configuration to a spell checker.
+  possible_err = new_aspell_speller(spell_config);
+  if (aspell_error_number(possible_err) != 0) {
+    // Failed on configuration.
+    FUZZ_DEBUG("Failed to create speller: %s",
+               aspell_error_message(possible_err));
+    delete_aspell_can_have_error(possible_err);
+    goto EXIT_LABEL;
+  }
+
+  // Create a spell checker.
+  spell_checker = to_aspell_speller(possible_err);
+
+  // Convert the spell checker to a document checker.
+  doc_err = new_aspell_document_checker(spell_checker);
+  if (aspell_error(doc_err) != 0) {
+    // Failed to convert to a document checker.
+    FUZZ_DEBUG("Failed to create document checker: %s",
+               aspell_error_message(doc_err));
+    delete_aspell_can_have_error(doc_err);
+    goto EXIT_LABEL;
+  }
+
+  doc_checker = to_aspell_document_checker(doc_err);
+
+  // Process the remainder of the document.
+  aspell_document_checker_process(doc_checker, data_str, size);
+
+  // Iterate over all misspellings.
+  token = aspell_document_checker_next_misspelling(doc_checker);
+
+  FUZZ_DEBUG("Token len %d", token.len);
+
+  for (;
+       token.len != 0;
+       token = aspell_document_checker_next_misspelling(doc_checker))
+  {
+    // Get spelling suggestions for the misspelling.
+    auto word_list = aspell_speller_suggest(spell_checker,
+                                            data_str + token.offset,
+                                            token.len);
+
+    // Iterate over the suggested replacement words in the word list.
+    AspellStringEnumeration *els = aspell_word_list_elements(word_list);
+
+    for (const char *word = aspell_string_enumeration_next(els);
+         word != 0;
+         word = aspell_string_enumeration_next(els))
+    {
+      // Conditionally print out the suggested replacement words.
+      FUZZ_DEBUG("Suggesting replacement for word at offset %d len %d: %s",
+                 token.offset,
+                 token.len,
+                 word);
+    }
+    delete_aspell_string_enumeration(els);
+  }
+
+EXIT_LABEL:
+
+  if (doc_checker != NULL) {
+    delete_aspell_document_checker(doc_checker);
+  }
+
+  if (spell_checker != NULL) {
+    delete_aspell_speller(spell_checker);
+  }
+
+  if (spell_config != NULL) {
+    delete_aspell_config(spell_config);
+  }
+
+  return 0;
+}
+
+// Returns -1 on error, or the number of bytes consumed from the config string
+// otherwise.
+int parse_config(AspellConfig *spell_config,
+                 uint8_t *config,
+                 size_t config_len)
+{
+  uint8_t line[MAX_CONFIG_LEN];
+
+  uint8_t *config_ptr = config;
+  size_t config_ptr_used = 0;
+
+  uint8_t *delimiter;
+
+  // Iterate over the lines.
+  for (delimiter = (uint8_t *)memchr(config_ptr,
```

Comment:
I would use `strchr` here.  If the fuzzier inserted a null character, it's not going to lead to meaningful results anyway.  Using memchr elsewhere is fine as we know there can't possible be a null character in the string.

Reply:
Disagree that it's not going to get meaningful results. I specifically used memchr here because having weird variables being input into the API calls is precisely the point of the fuzzer! Now realistically I know having null characters in weird places probably isn't going to cause many problems, but there's really no harm here in being permissive. In fact it's probably better.